### DISCUSSION TITLE
Define methods for instances with populated relations

### DISCUSSION THREAD

**Discussion by JulianCissen** [mikro-orm/mikro-orm#4500] [Q&A]
Hi,

I'm doing some changes around my codebase. One thing is that I'm trying to improve the usage of OOP. One of the things I'm doing to achieve this is take some scattered functions and relocate them to the class they belong to. Sometimes these functions require an entity to have a specific relation populated (done by setting the parameter type to `Loaded<Entity, 'relation'>`).

Now I want to functionally achieve the same using OOP. In this case that would mean that the method should only appear on an entity that has specific relations loaded. Is it possible to achieve this currently?

I can already define a synchronous method that throws if the collection isn't initialized, but I want to prevent alltogether that this situation can occur. It'd be better if the method just didn't exist on an entity that doesn't have the relation loaded.

A code block to illustrate my point:
```ts
import { Entity, Property, PrimaryKey, OneToMany, ManyToOne, Collection, Ref } from "@mikro-orm/core";
import { MikroORM } from "@mikro-orm/postgresql";

@Entity()
export class Author {
  @PrimaryKey()
  id!: number;

  @Property()
  name!: string;

  @OneToMany({
    entity: () => Book,
    mappedBy: 'author'
  })
  books = new Collection<Book>(this);

  // This function should only be available if the entity type is Loaded<Author, 'books'>.
  // Or any loaded type that indicates book is loaded.
  public bookTags() {
    return this.books.getItems().map((book) => book.tag);
  }
}

@Entity()
export class Book {
  @PrimaryKey()
  id!: number;

  @Property()
  name!: string;

  @Property()
  tag?: string;

  @ManyToOne({
    entity: () => Author,
    inversedBy: 'books'
  })
  author!: Ref<Author>;
}

const start = async () => {
  const db = await MikroORM.init({
      entities: [Book, Author],
      type: 'postgresql',
      dbName: 'debug',
      host: 'localhost',
      port: 5432,
      user: 'postgres',
      password: 'postgres',
      debug: true,
  });

  let em = db.em.fork();

  const author = await em.findOne(Author, { id: 1 });
  if (!author) throw new Error();
  // Should not be available at this point.
  author.bookTags();

  const anotherAuthor = await em.findOne(Author, { id: 2 }, { populate: ['books'] });
  if (!anotherAuthor) throw new Error();
  // Should be available since type is Loaded<Author, 'books'>.
  anotherAuthor.bookTags();
}

start();
```

**Comment by B4nan** [comment]
Nope, that's probably not possible, and I am not sure how could we implement it now, but it's an interesting idea.
